### ISSUE TITLE
💅 `complexity/nouselessconstructor` false positive

### ISSUE THREAD

**Issue by jpike88** [biomejs/biome#7495]
### Environment information

```bash
CLI:
  Version:                      2.2.4
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v24.0.1
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         npm/11.5.2

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            false
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  false

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  unset
  Enabled rules:
    a11y/noAccessKey
    a11y/noAriaHiddenOnFocusable
    a11y/noAriaUnsupportedElements
    a11y/noAutofocus
    a11y/noDistractingElements
    a11y/noHeaderScope
    a11y/noInteractiveElementToNoninteractiveRole
    a11y/noLabelWithoutControl
    a11y/noNoninteractiveElementToInteractiveRole
    a11y/noNoninteractiveTabindex
    a11y/noPositiveTabindex
    a11y/noRedundantAlt
    a11y/noRedundantRoles
    a11y/noStaticElementInteractions
    a11y/noSvgWithoutTitle
    a11y/useAltText
    a11y/useAnchorContent
    a11y/useAriaActivedescendantWithTabindex
    a11y/useAriaPropsForRole
    a11y/useAriaPropsSupportedByRole
    a11y/useButtonType
    a11y/useFocusableInteractive
    a11y/useGenericFontNames
    a11y/useHeadingContent
    a11y/useHtmlLang
    a11y/useIframeTitle
    a11y/useKeyWithClickEvents
    a11y/useKeyWithMouseEvents
    a11y/useMediaCaption
    a11y/useSemanticElements
    a11y/useValidAnchor
    a11y/useValidAriaProps
    a11y/useValidAriaRole
    a11y/useValidAriaValues
    a11y/useValidAutocomplete
    a11y/useValidLang
    complexity/noAdjacentSpacesInRegex
    complexity/noArguments
    complexity/noCommaOperator
    complexity/noEmptyTypeParameters
    complexity/noFlatMapIdentity
    complexity/noForEach
    complexity/noStaticOnlyClass
    complexity/noThisInStatic
    complexity/noUselessCatch
    complexity/noUselessConstructor
    complexity/noUselessContinue
    complexity/noUselessEmptyExport
    complexity/noUselessEscapeInRegex
    complexity/noUselessFragments
    complexity/noUselessLabel
    complexity/noUselessLoneBlockStatements
    complexity/noUselessRename
    complexity/noUselessStringRaw
    complexity/noUselessSwitchCase
    complexity/noUselessTernary
    complexity/noUselessThisAlias
    complexity/noUselessTypeConstraint
    complexity/noUselessUndefinedInitialization
    complexity/useDateNow
    complexity/useFlatMap
    complexity/useIndexOf
    complexity/useLiteralKeys
    complexity/useNumericLiterals
    complexity/useOptionalChain
    correctness/noChildrenProp
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noConstantMathMinMaxClamp
    correctness/noConstructorReturn
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalObjectCalls
    correctness/noInnerDeclarations
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noInvalidDirectionInLinearGradient
    correctness/noInvalidGridAreas
    correctness/noInvalidPositionAtImportRule
    correctness/noInvalidUseBeforeDeclaration
    correctness/noMissingVarFunction
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noSetterReturn
    correctness/noStringCaseMismatch
    correctness/noUnknownFunction
    correctness/noUnknownMediaFeatureName
    correctness/noUnknownProperty
    correctness/noUnknownPseudoClass
    correctness/noUnknownPseudoElement
    correctness/noUnknownTypeSelector
    correctness/noUnknownUnit
    correctness/noUnmatchableAnbSelector
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedFunctionParameters
    correctness/noUnusedImports
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/noVoidElementsWithChildren
    correctness/noVoidTypeReturn
    correctness/useGraphqlNamedOperations
    correctness/useIsNan
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    nursery/noFloatingPromises
    performance/noAccumulatingSpread
    performance/noDynamicNamespaceImportAccess
    security/noBlankTarget
    security/noDangerouslySetInnerHtml
    security/noDangerouslySetInnerHtmlWithChildren
    security/noGlobalEval
    style/noDefaultExport
    style/noNonNullAssertion
    style/noUnusedTemplateLiteral
    style/useArrayLiterals
    style/useAsConstAssertion
    style/useConsistentArrayType
    style/useConst
    style/useDeprecatedReason
    style/useExponentiationOperator
    style/useExportType
    style/useForOf
    style/useImportType
    style/useLiteralEnumMembers
    style/useNamingConvention
    style/useNodejsImportProtocol
    style/useNumberNamespace
    style/useObjectSpread
    style/useSelfClosingElements
    style/useShorthandFunctionType
    style/useSingleVarDeclarator
    suspicious/noApproximativeNumericConstant
    suspicious/noArrayIndexKey
    suspicious/noAssignInExpressions
    suspicious/noBiomeFirstException
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCommentText
    suspicious/noCompareNegZero
    suspicious/noConfusingLabels
    suspicious/noConfusingVoidType
    suspicious/noConsole
    suspicious/noConstEnum
    suspicious/noControlCharactersInRegex
    suspicious/noDebugger
    suspicious/noDocumentCookie
    suspicious/noDoubleEquals
    suspicious/noDuplicateAtImportRules
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateCustomProperties
    suspicious/noDuplicateElseIf
    suspicious/noDuplicateFields
    suspicious/noDuplicateFontNames
    suspicious/noDuplicateJsxProps
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noDuplicateProperties
    suspicious/noDuplicateSelectorsKeyframeBlock
    suspicious/noEmptyBlock
    suspicious/noEmptyInterface
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noGlobalIsFinite
    suspicious/noImplicitAnyLet
    suspicious/noImportAssign
    suspicious/noImportantInKeyframe
    suspicious/noIrregularWhitespace
    suspicious/noLabelVar
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noMisrefactoredShorthandAssign
    suspicious/noOctalEscape
    suspicious/noPrototypeBuiltins
    suspicious/noQuickfixBiome
    suspicious/noRedundantUseStrict
    suspicious/noSelfCompare
    suspicious/noShorthandPropertyOverrides
    suspicious/noSparseArray
    suspicious/noSuspiciousSemicolonInJsx
    suspicious/noThenProperty
    suspicious/noTsIgnore
    suspicious/noUnassignedVariables
    suspicious/noUnknownAtRules
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noUselessRegexBackrefs
    suspicious/noVar
    suspicious/noWith
    suspicious/useAdjacentOverloadSignatures
    suspicious/useBiomeIgnoreFolder
    suspicious/useDefaultSwitchClauseLast
    suspicious/useGetterReturn
    suspicious/useGoogleFontDisplay
    suspicious/useIsArray
    suspicious/useIterableCallbackReturn
    suspicious/useNamespaceKeyword

Workspace:
  Open Documents:               0
```

### Rule name

complexity/nouselessconstructor

### Playground link

https://biomejs.dev/playground/?lintRules=all&code=aQBuAHQAZQByAGYAYQBjAGUAIABBACAAewB4ADoAbgB1AG0AYgBlAHIAfQA7AAoAaQBuAHQAZQByAGYAYQBjAGUAIABCACAAewB5ADoAcwB0AHIAaQBuAGcAfQA7AAoACgBjAGwAYQBzAHMAIABNAHkAQgBhAHMAZQBDAGwAYQBzAHMAIAB7AAoACgAgACAAYwBvAG4AcwB0AHIAdQBjAHQAbwByACgAYQByAGcAOgBBAHwAQgApAHsACgAgACAAIAAgAGEAcgBnADsACgAgACAAfQAKACAAIAAKAH0AOwAKAAoAYwBsAGEAcwBzACAATQB5AFMAdQBiAEMAbABhAHMAcwAgAGUAeAB0AGUAbgBkAHMAIABNAHkAQgBhAHMAZQBDAGwAYQBzAHMAIAB7AAoACgAgACAAYwBvAG4AcwB0AHIAdQBjAHQAbwByACgAYQByAGcAOgBBACkAewAKACAAIAAgACAAcwB1AHAAZQByACgAYQByAGcAKQA7AAoAIAAgAH0ACgAgACAACgB9AAoACgBuAGUAdwAgAE0AeQBTAHUAYgBDAGwAYQBzAHMAKAB7AHgAOgAzAH0AKQA%3D

### Expected result

The rule shouldn't complain, as the constructor is necessary to enforce the narrowed type of that subclass. I have a base class with over 50 subclasses that use it in this way, none of the valid options for the rule are options (default values would be time consuming and weird to create, can't make the property private as it's supposed to be public, etc)

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
>  The rule shouldn't complain

Can we come up with the specific criteria of when the rule shouldn't complain? Also, how does TypeScript eslint behave in this case?

**Comment by siketyan** [comment]
It seems typescript-eslint complains too: https://typescript-eslint.io/play#ts=5.9.2&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgILIN4A8BcICuAtgEbQC%2BA3AFCiSyIoBCmAnjgM5hSgDmlVVBABs47dsgCyLRqIgBhEWMwDkyBAHsQnKPgRh1UABRwoPHKgA%2BjAJQYqq1SZ7VVZe8ir9Bi8VIDK%2BMQKouIQWJAgACa%2B0rLBSnbuGlpcuvpGTua27qrs%2BAAO0Mam1i7IbqqeAiAQAO6SLAFBPoYYyLjIAMzl1lRAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tmUSWWS0yHJsnxwy%2BDtHRRE0aNMjgwAXxCqgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

**Comment by jpike88** [comment]
> It seems typescript-eslint complains too

In this particular case I wouldn't mirror their situation, I believe they couldn't do it because it's a limitation of the tool itself

https://typescript-eslint.io/rules/no-useless-constructor/#caveat
https://github.com/typescript-eslint/typescript-eslint/issues/9884#issuecomment-2310076869

**Comment by jpike88** [comment]
The rule shouldn't complain, if:

- super is being called in the constructor
- at least one parameter is being passed to super
- some kind of judgement is made types-wise* (to be 100% correct, the child class params are types that fall into the parent param's types)

For the type comparison bit, that may be a little overkill as Typescript itself is there to catch type misalignments like that. You could just do a simple 1:1 comparison and if there is any difference at all, just ignore it. I don't know exactly what strategy is best at that point, you're the genius on that. You might decide that anything being passed to super is enough for the rule to ignore it

**Comment by vladimir-ivanov** [comment]
@arendjr would graph module be able to read parent class constructor param types please? and make a match on that? If so I may look into this.

**Comment by Conaclos** [comment]
I think this was already reported in the past. We decided to not address the issue because the constructor is still useless from a JavaScript perspective: types are erased at compilation.

However, I see the inconvenience. And because it is brought a second time, perhaps it is worth to consider. I see two possible approaches: ignoring this constructor that pass an argument to `super` in TypeScript files only, or relying on our type service to detect this overriding pattern. In the second case it could be nice to only pull the service in TypeScript files; However it is not currently possible.

**Comment by arendjr** [comment]
Yeah, we can indeed do a lookup for the parent class, and silence the diagnostic in case the types of the arguments are different from those defined in the parent class.

For what it's worth, I think the pattern may apply to JavaScript files too, even though they don't have explicit type annotations. Consider this:

```js
class Base {
    constructor(foo, bar) {
        // do stuff
    }
}

class Sub extends Base {
    constructor(foo) {
        super(foo)
    }
}
```

In addition, this TypeScript-specific situation should also be silenced, even though the arguments have no difference in types:

```ts
class Base {
    protected constructor(foo: Foo) {
        // do stuff
    }
}

class Sub extends Base {
    constructor(foo: Foo) {
        super(foo)
    }
}
```

This one is currently listed as a known caveat, but if we implement lookups for the parent class, I think it's best to address it too. Could be a separate improvement/PR, of course.
